### PR TITLE
Improve `LightningDataModule` hook test and fix `dataloader_idx` argument

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Create `HookedDataModule` to append the called hooks dynamically. Any new hook will make the test fail. We want this so there's no forgetting to update the test.
- The hook arguments are now tested
- A small fix is included to only include the `dataloader_idx` argument during predict if there is more than one dataloader. This is consistent with the rest of the loops.

Related to the Lightning hooks review: #7740

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGMT !

**Review by awaelchli — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -226,6 +226,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed a bug where `precision=64` with `accelerator='ddp_spawn'` would throw a pickle error ([#6924](https://github.com/PyTorchLightning/pytorch-lightning/pull/6924))
 
 
+- Fixed `dataloader_idx` argument value when predicting with only one `DataLoader` ([#7941](https://github.com/PyTorchLightning/pytorch-lightning/pull/7941))
+
+
+- Fixed broadcasting in multi-node, multi-gpu DDP using torch 1.7 ([#7592](https://github.com/PyTorchLightning/pytorch-lightning/pull/7592))
+
+
```
Comment (carmocca) [inline CHANGELOG.md:234]:
```suggestion
```

**Review by justusschock — APPROVED** [review]
